### PR TITLE
Unify docs back to .py files

### LINKED ISSUE #40: Restore docs from .rst to docstrings?
Back in https://github.com/zopefoundation/zope.configuration/commit/941b1c84f2a2477116a70ae19a56d96a59fc630e , among others, during the  Python 3 porting process, lots (but not all) of the documentation for modules, classes and functions was removed from docstrings and put into the .rst files. The reason was that the docstrings could only be tested under Python 2.

This makes it difficult to know where to document something, or if something even is documented. It might not have a docstring, but still be explicitly documented in sphinx.

Now that the Sphinx doctests have been ported to run on both python 2 and 3 (#24), can we re-unify all the documentation, back into the docstrings? The .rst files would then just consist of some `.. automodule::` or at worst `.. autoclass::` statements.

### PR DESCRIPTION
**Pull request by jamadden**
Fixes #40

### CODE REVIEW

File: src/zope/configuration/config.py [hunk 1]
```diff
@@ -828,6 +1562,98 @@ def toargs(context, schema, data):
...
+        >>> class schema(Interface):
+        ...     in_ = Int(constraint=lambda v: v > 0)
+        ...     f = Float()
+        ...     n = TextLine(min_length=1, default=u"rob")
+        ...     x = BytesLine(required=False)
+        ...     u = URI()
+
+        >>> context = ConfigurationMachine()
+        >>> from pprint import PrettyPrinter
+        >>> pprint=PrettyPrinter(width=50).pprint
```
Comment (mgedmin) [inline src/zope/configuration/config.py]: Spaces around the assignment `=`?

File: src/zope/configuration/xmlconfig.py [hunk 2]
```diff
@@ -100,10 +117,40 @@ def __str__(self):
...
+    """
+    Information about a directive based on parser data
 
     This includes the directive location, as well as text data
     contained in the directive.
+
+    Example
+
+        >>> from zope.configuration.xmlconfig import ParserInfo
+        >>> info = ParserInfo('tests//sample.zcml', 1, 0)
```
Comment (mgedmin) [inline src/zope/configuration/xmlconfig.py]: The doubled slashes confuse me.

(I'm sorry, I don't know how to review patches that *move code around*, all I see is newly-added code!)

File: src/zope/configuration/xmlconfig.py [hunk 3]
```diff
@@ -100,10 +117,40 @@ def __str__(self):
...
+        >>> info
+        File "tests//sample.zcml", line 1.0
+
+        >>> print(info)
+        File "tests//sample.zcml", line 1.0
+
+        >>> info.characters("blah\\n")
```
Comment (mgedmin) [inline src/zope/configuration/xmlconfig.py]: Raw docstrings FTW

File: src/zope/configuration/xmlconfig.py [hunk 4]
```diff
@@ -173,7 +220,8 @@ def characters(self, characters):
 
 
 class ConfigurationHandler(ContentHandler):
-    """Interface to the xml parser
+    """
+    Interface to the xml parser
```
Comment (mgedmin) [inline src/zope/configuration/xmlconfig.py]: Capitalize XML?

File: src/zope/configuration/fields.py [hunk 5]
```diff
@@ -195,15 +430,99 @@ def fromUnicode(self, value):
...
+          >>> context = FauxContext()
+          >>> field = MessageID().bind(context)
+
+        There is a fallback domain when no domain has been specified.
+
+        Exchange the warn function so we can make test whether the warning
+        has been issued
+
+          >>> warned = None
+          >>> def fakewarn(*args, **kw): #* syntax highlighting
```
[the remark this answers is not in the document]
Comment (jamadden) [reply src/zope/configuration/fields.py]: It doesn't seem to make any difference.

File: src/zope/configuration/tests/test_docs.py [hunk 6]
```diff
@@ -84,4 +97,15 @@ def test_suite():
         )
     )
 
+    for mod_name in api_to_test:
+        mod_name = 'zope.configuration.' + mod_name
+        __import__(mod_name)
+        suite.addTest(
+            doctest.DocTestSuite(
+                sys.modules[mod_name],
```
[the remark this answers is not in the document]
Comment (jamadden) [reply src/zope/configuration/tests/test_docs.py]: Ah, thanks. I don't use DocTestSuite very often.

File: src/zope/configuration/config.py [hunk 7]
```diff
@@ -265,13 +371,114 @@ def processFile(self, filename):
...
+             {'args': (),
+              'callable': None,
+              'discriminator': None,
+              'includepath': ('foo.zcml',),
+              'info': '?',
+              'kw': {},
+              'order': 99999}
+
+        We can also pass an includepath argument, which will be used as the the
+        includepath for the action.  (if includepath is None, self.includepath
```
[the remark this answers is not in the document]
Comment (jamadden) [reply src/zope/configuration/config.py]: includepath is important for resolving conflicts.

File: src/zope/configuration/config.py [hunk 8]
```diff
@@ -299,24 +506,91 @@ def action(self, discriminator, callable=None, args=(), kw=None, order=0,
...
+
+    Examples:
+
+        >>> from zope.configuration.interfaces import IConfigurationContext
+        >>> from zope.configuration.config import ConfigurationAdapterRegistry
+        >>> from zope.configuration.config import ConfigurationError
+        >>> from zope.configuration.config import ConfigurationMachine
+        >>> r = ConfigurationAdapterRegistry()
+        >>> c = ConfigurationMachine()
+        >>> r.factory(c, ('http://www.zope.com','xxx'))
```
[the remark this answers is not in the document]
Comment (jamadden) [reply src/zope/configuration/config.py]: That's the kind of thing my linter (pylint) usually warns about, but it doesn't do so inside doctests in docstrings. I wonder if there's a way to make that happen.

File: src/zope/configuration/config.py [hunk 9]
```diff
@@ -402,9 +705,56 @@ def setInfo(self, info):
...
+        """
+        Execute the configuration actions.
 
         This calls the action callables after resolving conflicts.
+
+        For example:
+
+            >>> from zope.configuration.config import ConfigurationMachine
+            >>> output = []
+            >>> def f(*a, **k): #* syntax highlighting
```
[the remark this answers is not in the document]
Comment (jamadden) [reply src/zope/configuration/config.py]: I think it's more about docutils/sphinx. Even now, the highlighting for `fields.Bool.fromUnicode` is broken (and doing this kind of thing doesn't help it).

File: src/zope/configuration/config.py [hunk 10]
```diff
@@ -571,14 +1070,135 @@ def noop():
...
+    a method for each subdirective:
+
+        >>> class Handler(object):
+        ...   def __init__(self, context, x, y):
+        ...      self.context, self.x, self.y = context, x, y
+        ...      context.action('init', f)
+        ...   def sub(self, context, a, b):
+        ...      context.action(('sub', a, b), f)
+        ...   def __call__(self):
+        ...      self.context.action(('call', self.x, self.y), f)
```
[the remark this answers is not in the document]
Comment (jamadden) [reply src/zope/configuration/config.py]: I think I'll leave that kind of whitespace alone for now. My editor doesn't auto-indent the doctests inside docstrings, so I have to count spaces, which is annoying at best.

**Comment by jamadden** [comment]
Thanks! I incorporated most of your (simpler) feedback.

File: src/zope/configuration/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mgedmin) [reply src/zope/configuration/config.py]: My linter (flake8) can lint code in doctests because I submitted a PR for it to pyflakes ages ago ;)

**Pull request merged**